internal/cloud/backend.go

```diff
@@ -437,6 +439,51 @@ func (b *Cloud) retryLogHook(attemptNum int, resp *http.Response) {
 	}
 }
 
+func (b *Cloud) Workspace(name string) (*tfe.Workspace, error) {
+	workspace, err := b.client.Workspaces.Read(context.Background(), b.organization, name)
+	if err != nil {
+		return nil, err
+	}
+	return workspace, nil
+}
+
+type outputData struct {
+	Value     cty.Value
+	Sensitive bool
+}
+type StateOutputData struct {
+	Outputs map[string]*outputData
+}
+
+func (b *Cloud) StateVersionOutputs(svID string) (*StateOutputData, error) {
+	svo, err := b.client.StateVersions.Outputs(context.Background(), svID, tfe.StateVersionOutputsListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	sd := &StateOutputData{
```

Comment:
I was stumped on how to convert from the go types to cty types. `cty.ImpliedType` takes in `interface{}`, which is what go-tfe uses for the Value, but for some reason cty does not like the `interface{}` that it receives. I asked Omar, and he showed me where he solved this problem elsewhere: https://github.com/hashicorp/terraform-provider-tfe/blob/main/tfe/data_source_outputs.go#L165-L189

Is there a better way to accomplish this transformation? /cc @alisdair, @apparentlymart, @chrisarcand

Reply:
In order to answer this question I tried to work my way back to where this data comes from and of course the real source is inside a not open source codebase and so I can't link to it here, but the closest layer we can see from outside is [the state version outputs endpoint](https://www.terraform.io/cloud-docs/api-docs/state-version-outputs#list-state-version-outputs), which I _believe_ (but am not 100% sure) has the same structure as the embedded output values that you get when using this `include=outputs` option.

Based on what I see in those API docs, it looks like the data we're getting from this API has already been transformed as compared to the raw value Terraform would normally store in a state snapshot, and so I suspect that it won't be possible to directly achieve the same result we would get from decoding the state: the API is discarding some details that our normal decoding relies on.

For example, I see in the example it says `"type": "array"`, which I have to assume is the result of some interpretation done inside Terraform Cloud because Terraform Core itself has no type called "array" -- it would either be a list type or a tuple type, which would look something like one of the following examples:

```json
["list", "string"]
```

```json
["tuple", ["string", "string", "string", "string"]]
```

I think in order to implement this correctly (so it will behave exactly the same as it would with direct state access) we will probably need to extend the API with some way to get the data exactly how Terraform CLI records it in the state, preserving the exact type information and value that Terraform originally recorded, and then we can decode it [the same way we would for a full state snapshot](https://github.com/hashicorp/terraform/blob/0900c7e0bb9ab7428e875b80eb73b89a5d5efbfe/internal/states/statefile/version4.go#L261-L291).